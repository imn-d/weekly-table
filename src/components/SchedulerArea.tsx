import React, { FC, useEffect } from 'react';

import {
  useCanvas,
  useCustomTime,
  usePointerLock,
  usePopup,
  useScheduler,
  useTimeBlock,
} from '../providers';

/**
 * Time blocks drawing area component
 */
export const SchedulerArea: FC = () => {
  const { baseZIndex } = useScheduler();
  const { startDrawing, finishDrawing, draw, canvasRef, isDrawing } =
    useCanvas();
  const { blocks } = useTimeBlock();
  const { hidePopup } = usePopup();
  const { isLocking } = usePointerLock();
  const { isEditing } = useCustomTime();

  /**
   * Automatically hides the popup if at the same time:
   * no draws new block
   * no uses pointer lock
   * no custom time setter is active
   */
  useEffect(() => {
    if (!isDrawing && !isLocking && !isEditing) {
      hidePopup();
    }
    //eslint-disable-next-line
  }, [isDrawing, blocks, isLocking, isEditing]);

  return (
    <canvas
      id={'schedulerArea'}
      className={'scheduler-area'}
      style={{
        position: 'absolute',
        zIndex: baseZIndex + 3,
        cursor: 'crosshair',
      }}
      onMouseDown={(event) => (isEditing ? null : startDrawing(event))}
      onMouseUp={finishDrawing}
      onMouseMove={draw}
      onMouseLeave={finishDrawing}
      ref={canvasRef}
    />
  );
};
