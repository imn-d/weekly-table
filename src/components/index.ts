export * from './Background';
export * from './Popup';
export * from './TimeBlockCore';
export * from './CustomTime';
export * from './SchedulerArea';
export * from './SchedulerHeader';
export * from './SchedulerHelper';
export * from './SchedulerLayout';
