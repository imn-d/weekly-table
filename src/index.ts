import {
  BlockColorsProps,
  ScheduleGroup,
  SchedulerColumnsProps,
  SchedulerInputProps,
  schedulerColumns,
  schedulerRows,
} from './common';
import Scheduler from './providers/Scheduler';

export default Scheduler;
export { schedulerColumns, schedulerRows };
export type {
  SchedulerColumnsProps,
  SchedulerInputProps,
  ScheduleGroup,
  BlockColorsProps,
};
