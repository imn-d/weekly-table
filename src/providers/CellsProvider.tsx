import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { Cell, CellsProps, HOUR_24 } from '../common';
import { useScheduler } from './Scheduler';

const CellsContext = createContext<CellsProps>({} as CellsProps);
/**
 *  Virtual cells provider context hook
 *  @returns [CellsProps]{@link CellsProps}
 */
export const useCells = (): CellsProps => useContext(CellsContext);

/**
 * Virtual cells provider
 */
const CellsProvider: FC = ({ children }) => {
  const {
    parentRef,
    timeframe,
    columns,
    helperWidthProp,
    headerHeightProp,
    bottomHeightProp,
  } = useScheduler();

  /**
   * Main context state, setting by useEffect
   */
  const [cellsProps, setCellsProps] = useState<CellsProps>({} as CellsProps);

  /**
   * Main context updater
   */
  const setCells = useCallback(
    (update: Cell[]) => {
      setCellsProps({ ...cellsProps, cells: update });
    },
    [cellsProps],
  );

  /**
   * Ms count per timeframe
   */
  const msTime = useMemo(() => {
    return timeframe * 60 * 1000;
  }, [timeframe]);

  /**
   * Calculates context values and build virtual cells area
   */
  useEffect(() => {
    const parent = parentRef?.current;
    if (!parent) return;
    const widthStep = (parent.offsetWidth - helperWidthProp) / columns?.length;
    const areaHeight =
      parent.offsetHeight - headerHeightProp - bottomHeightProp;
    const heightStep = areaHeight / (1440 / timeframe);
    const millisPerPixel = HOUR_24 / areaHeight;
    const maxCellIndex = areaHeight / heightStep - 1;
    const cells = new Array<Cell>();
    let widthPos = helperWidthProp;
    for (let i = 0; i < columns?.length; i++) {
      let heightPos = headerHeightProp;
      for (let j = 0; j < areaHeight / heightStep; j++) {
        cells.push({
          isSelected: false,
          column: i,
          row: j,
          position: {
            x: widthPos,
            y: heightPos,
            w: widthPos + widthStep,
            h: heightPos + heightStep,
          },
        } as Cell);
        heightPos += heightStep;
      }
      widthPos += widthStep;
    }

    setCellsProps({
      widthStep,
      heightStep,
      millisPerPixel,
      cells,
      maxCellIndex,
      offsetWidth: parent.offsetWidth,
      offsetHeight: parent.offsetHeight,
    } as CellsProps);
  }, [
    parentRef?.current?.offsetWidth,
    parentRef?.current?.offsetHeight,
    columns?.length,
    bottomHeightProp,
    headerHeightProp,
    helperWidthProp,
    timeframe,
    parentRef,
  ]);

  return (
    <CellsContext.Provider
      value={{
        ...cellsProps,
        setCells: setCells,
        msTime,
      }}
    >
      {children}
    </CellsContext.Provider>
  );
};

export default CellsProvider;
