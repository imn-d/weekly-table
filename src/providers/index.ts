export * from './CanvasProvider';
export * from './CellsProvider';
export * from './CustomTimeProvider';
export * from './PointerLockProvider';
export * from './PopupProvider';
export * from './Scheduler';
export * from './TimeBlockProvider';
