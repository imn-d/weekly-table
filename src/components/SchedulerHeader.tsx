import React, { FC } from 'react';

import { useCells, useScheduler } from '../providers';

/**
 *  Days of week component (columns)
 */
export const SchedulerHeader: FC = () => {
  const { widthStep } = useCells();
  const { headerHeightProp, columns } = useScheduler();
  return (
    <div
      id={'schedulerHeader'}
      className={'scheduler-header'}
      style={{
        position: 'relative',
        userSelect: 'none',
        display: 'inline-flex',
        height: `${headerHeightProp}px`,
        maxHeight: `${headerHeightProp}px`,
      }}
    >
      {columns?.map((col) => {
        return (
          <div
            id={'schedulerHeaderColumn'}
            className={'scheduler-text scheduler-header-column'}
            style={{
              display: 'inline-flex',
              minWidth: `${widthStep}px`,
              width: `${widthStep}px`,
            }}
            key={col.weight}
          >
            {col.full}
          </div>
        );
      })}
    </div>
  );
};
