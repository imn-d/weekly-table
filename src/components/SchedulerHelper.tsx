import React, { FC, useMemo } from 'react';

import { useCells, useScheduler } from '../providers';

/**
 *  Time description component (first column)
 */
export const SchedulerHelper: FC = () => {
  const { helperWidthProp, rows, headerHeightProp, bottomHeightProp } =
    useScheduler();
  const { offsetHeight } = useCells();

  /**
   * Calculates height for time descriptions assignment
   */
  const helperHeight = useMemo(() => {
    return (offsetHeight - headerHeightProp - bottomHeightProp) / rows?.length;
  }, [bottomHeightProp, headerHeightProp, offsetHeight, rows?.length]);

  return (
    <div
      id={'schedulerHelper'}
      className={'scheduler-helper'}
      style={{
        display: 'inline-flex',
        flexDirection: 'column',
        position: 'relative',
        width: `${helperWidthProp}px`,
        maxWidth: `${helperWidthProp}px`,
        userSelect: 'none',
      }}
    >
      {rows?.map((row, index) => {
        return (
          <div
            id={'schedulerHelperRow'}
            className={'scheduler-text scheduler-helper-row'}
            key={index}
            style={{
              display: 'inline-flex',
              minHeight: `${helperHeight}px`,
              height: `${helperHeight}px`,
            }}
          >
            {row}
          </div>
        );
      })}
    </div>
  );
};
