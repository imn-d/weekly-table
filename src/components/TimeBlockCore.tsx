import React, { FC, useCallback, useMemo } from 'react';

import { HOUR_24, TimeBlock } from '../common';
import { useCanvas, useCustomTime, usePopup, useScheduler } from '../providers';

/**
 * Each time block represents by this component
 * @param block - [TimeBlock]{@link TimeBlock}
 */
export const TimeBlockCore: FC<TimeBlock> = ({ ...block }: TimeBlock) => {
  const { baseZIndex, blockColors } = useScheduler();
  const { isDrawing } = useCanvas();
  const { popup, showPopup, hidePopup } = usePopup();
  const { isEditing, handleObject } = useCustomTime();

  /**
   * Select blocks colors
   */
  const bgColor = useMemo<string>(() => {
    if (popup?.id === block.id) return blockColors.hover;
    if (block.isTemp) return blockColors.temp;
    if (isDrawing) return blockColors.draw;
    return blockColors.common;
  }, [
    block.id,
    block.isTemp,
    blockColors.common,
    blockColors.draw,
    blockColors.hover,
    blockColors.temp,
    isDrawing,
    popup?.id,
  ]);

  /**
   * Calling when time block is hovering, popup and custom time components setter
   */
  const handleShowActions = useCallback(() => {
    if (isEditing) return null;
    showPopup(block.id);
    handleObject(block);
    //eslint-disable-next-line
  }, [block, isEditing]);

  /**
   * Showing endTime as 24:00 instead of 00:00
   */
  const endTime = useMemo<string>(() => {
    if (block.endTime === HOUR_24) return '24:00';
    return new Date(block.endTime).toISOString().substr(11, 5);
  }, [block.endTime]);

  return (
    <div
      id={'timeBlock'}
      className={'scheduler-time-block'}
      style={{
        zIndex: isDrawing ? baseZIndex : baseZIndex + 4,
        position: 'absolute',
        top: block.top,
        left: block.left,
        width: block.width,
        height: block.height,
        backgroundColor: bgColor,
      }}
      onMouseEnter={handleShowActions}
      onMouseMove={handleShowActions}
      onMouseLeave={() => (!isEditing ? hidePopup() : null)}
    >
      <span
        id={'timeDisplay'}
        className={'scheduler-text scheduler-time-display'}
        style={{
          userSelect: 'none',
          pointerEvents: 'none',
        }}
      >
        {new Date(block.startTime).toISOString().substr(11, 5) +
          ' / ' +
          endTime}
      </span>
    </div>
  );
};
