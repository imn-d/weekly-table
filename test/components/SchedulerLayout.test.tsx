import { mount } from 'enzyme';
import React from 'react';

import {
  CellsProps,
  SchedulerProps,
  TimeBlock,
  TimeBlockProps,
} from '../../src/common';
import { SchedulerLayout } from '../../src/components';
import * as scArea from '../../src/components/SchedulerArea';
import * as scHeader from '../../src/components/SchedulerHeader';
import * as scHelper from '../../src/components/SchedulerHelper';
import * as tBlock from '../../src/components/TimeBlockCore';
import * as uc from '../../src/providers/CellsProvider';
import * as us from '../../src/providers/Scheduler';
import * as ut from '../../src/providers/TimeBlockProvider';

describe('SchedulerLayout Component', () => {
  const useSchedulerMock = jest.spyOn(us, 'useScheduler');
  const useCellMock = jest.spyOn(uc, 'useCells');
  const useTimeMock = jest.spyOn(ut, 'useTimeBlock');
  const schedulerArea = jest.spyOn(scArea, 'SchedulerArea');
  const schedulerHeader = jest.spyOn(scHeader, 'SchedulerHeader');
  const schedulerHelper = jest.spyOn(scHelper, 'SchedulerHelper');
  const schedulerBlock = jest.spyOn(tBlock, 'TimeBlockCore');

  beforeEach(() => {
    jest.clearAllMocks();

    useSchedulerMock.mockReturnValue({
      headerHeightProp: 80,
      helperWidthProp: 80,
      bottomHeightProp: 20,
      baseZIndex: 0,
    } as SchedulerProps);

    useCellMock.mockReturnValue({
      offsetHeight: 100,
      offsetWidth: 200,
    } as CellsProps);

    schedulerArea.mockReturnValue(<div id={'area'}>Area</div>);
    schedulerHeader.mockReturnValue(<div id={'header'}>Header</div>);
    schedulerHelper.mockReturnValue(<div id={'helper'}>Helper</div>);
    schedulerBlock.mockReturnValue(<div id={'block'}>Block</div>);
  });

  it('Base Render', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    useTimeMock.mockReturnValueOnce({
      blocks: undefined,
      preview: undefined,
    } as TimeBlockProps);

    const wrapper = mount(<SchedulerLayout />);

    expect(useSchedulerMock).toBeCalledTimes(1);
    expect(useCellMock).toBeCalledTimes(1);
    expect(useTimeMock).toBeCalledTimes(1);

    expect(wrapper).toMatchSnapshot('baseRender');
    expect(wrapper.find('#area').hostNodes()).toHaveLength(1);
    expect(wrapper.find('#header').hostNodes()).toHaveLength(1);
    expect(wrapper.find('#helper').hostNodes()).toHaveLength(1);
    expect(wrapper.find('#block').hostNodes()).toHaveLength(0);
    wrapper.unmount();
  });

  it('Time Blocks Count Test', () => {
    useTimeMock.mockReturnValueOnce({
      blocks: new Array<TimeBlock>(
        { id: '1', isTemp: false } as TimeBlock,
        { id: '2', isTemp: false } as TimeBlock,
      ),
      preview: new Array<TimeBlock>(
        { id: '1', isTemp: true } as TimeBlock,
        { id: '2', isTemp: true } as TimeBlock,
      ),
    } as TimeBlockProps);

    const wrapper = mount(<SchedulerLayout />);

    const blocks = wrapper.find('#block').hostNodes();
    expect(blocks).toHaveLength(4);
    wrapper.unmount();
  });
});
