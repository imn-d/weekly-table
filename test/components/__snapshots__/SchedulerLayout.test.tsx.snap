// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchedulerLayout Component Base Render: baseRender 1`] = `
<SchedulerLayout>
  <div
    className="scheduler-layout"
    id="schedulerLayout"
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "80px 120px",
        "gridTemplateRows": "80px 0px 20px",
        "height": 100,
        "maxHeight": 100,
        "maxWidth": 200,
        "position": "absolute",
        "width": 200,
        "zIndex": 2,
      }
    }
  >
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    />
    <mockConstructor>
      <div
        id="header"
      >
        Header
      </div>
    </mockConstructor>
    <mockConstructor>
      <div
        id="helper"
      >
        Helper
      </div>
    </mockConstructor>
    <mockConstructor>
      <div
        id="area"
      >
        Area
      </div>
    </mockConstructor>
  </div>
</SchedulerLayout>
`;
